src/Hl7.Fhir.Base/Utility/ExtendedCodedException.cs

```diff
@@ -0,0 +1,84 @@
+﻿/* 
+ * Copyright (c) 2021, Firely (info@fire.ly) and contributors
+ * See the file CONTRIBUTORS for details.
+ * 
+ * This file is licensed under the BSD 3-Clause license
+ * available at https://raw.githubusercontent.com/FirelyTeam/firely-net-sdk/master/LICENSE
+ */
+
+using Hl7.Fhir.Model;
+using System;
+
+#nullable enable
+
+namespace Hl7.Fhir.Utility
+{
+    /// <summary>
+    /// This is a class of Exceptions that is raised by the SDK and is coded with a unique code
+    /// to enable callers to identify this exception and react appropriately on the code.
+    /// </summary>
+    /// <remarks>Most modules of the SDK using this Exception will create specific subclasses
+    /// for this subclass, providing a list of codes used by that module.</remarks>
+    public class ExtendedCodedException : CodedException
+    {
+        public ExtendedCodedException(string errorCode, string message, OperationOutcome.IssueSeverity issueSeverity, OperationOutcome.IssueType issueType) : base(errorCode, message)
+        {
+            IssueSeverity = issueSeverity;
+            IssueType = issueType;
+        }
+
+        public ExtendedCodedException(string errorCode, string message, OperationOutcome.IssueSeverity issueSeverity, OperationOutcome.IssueType issueType, Exception? innerException) : base(errorCode, message, innerException)
+        {
+            IssueSeverity = issueSeverity;
+            IssueType = issueType;
+        }
+
+        internal static string FormatLocationMessage(string baseMessage, string? instancePath, long? lineNumber, long? position)
+        {
+            string location = $"At line {lineNumber}, position {position}";
+            if (!string.IsNullOrEmpty(instancePath))
+                location = $"At {instancePath}, line {lineNumber}, position {position}";
+            var messageWithLocation = $"{baseMessage} {location}";
+            return messageWithLocation;
+        }
+
+        /// <summary>
+        /// Severity of this specific issue.
+        /// </summary>
+        /// <remarks>
+        /// Setter is public to permit others to upgrade/downgrade specific issues
+        /// as needed.
+        /// </remarks>
+        public OperationOutcome.IssueSeverity IssueSeverity { get; init; } = OperationOutcome.IssueSeverity.Error;
+
+        /// <summary>
+        /// Type of issue to report in a FHIR OperationOutcome.
+        /// </summary>
+        public OperationOutcome.IssueType IssueType { get; init; } = OperationOutcome.IssueType.Invalid;
+
+        /// <summary>
+        /// The error message without any location information appended to it (which is in Exception.Message property).
+        /// </summary>
+        public string? BaseErrorMessage { get; init; }
```

Comment:
The design of this property is incorrect. You can set this property on construction, but it will not change the exception at all. In fact, this message is supposed to be the base of the message you pass to the constructor (which includes location) etc, but that is not expressed anywhere structurally (it is in the comment).  This will for sure confuse the users.

I think we can improve on this: have the original constructors with just code  + message do the "old" behaviour (just for backwards compat, we'll not be using them anymore).  Then we add new constructors that explicitly take the code + *basemessage* + location + anything that influences the formatted message - that constructor will set the basemessage (if still necessary), construct the formatted message and pass that off to the base constructor.

This way you are either explicitly in backwards compat mode (and we can obsolete these constructors) OR in the new mode where you pass in the base message + the rest.

Reply:
I'll try to push a version that expresses what I mean.

Reply:
Done.